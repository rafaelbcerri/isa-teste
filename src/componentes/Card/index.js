import React, { useState } from 'react';
import PropTypes from 'prop-types';

function Card({
  title,
  text,
  onClick,
  buttonText = 'Enviar'
}) {
  const [ isTitleHidden, seIsTitleHidden ] = useState(false);
  const toggleHiddenTitle = () => seIsTitleHidden(!isTitleHidden);
  return (
    <div className='card'>
      {
        !isTitleHidden && (
          <>
            <h1>{title}</h1>
            <hr />
          </>
        )
      }
      <p>{text}</p>
      <button onClick={onClick}>
        {buttonText}
      </button>
      <button onClick={toggleHiddenTitle}>
        Esconde titulo
      </button>
    </div>
  )
}

Card.propTypes = {
  title: PropTypes.string.isRequired,
  text: PropTypes.string.isRequired,
  onClick: PropTypes.func.isRequired,
  buttonText: PropTypes.string
};

export default Card;