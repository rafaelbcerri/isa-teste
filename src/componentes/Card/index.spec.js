import {
  render,
  screen,
  fireEvent,
  act
} from '@testing-library/react';
import checkPropTypes from 'check-prop-types';

import Card from '.';

describe('Card', () => {
  let onClick = jest.fn();

  it('should check propTypes', () => {
    expect(
      checkPropTypes(
        { title: Card.propTypes.title },
        {},
        'prop',
        'Card',
      ),
    ).toBe(
      'Failed prop type: The prop `title` is marked as required in `Card`, but its value is `undefined`.',
    );
    expect(
      checkPropTypes(
        { text: Card.propTypes.text },
        {},
        'prop',
        'Card',
      ),
    ).toBe(
      'Failed prop type: The prop `text` is marked as required in `Card`, but its value is `undefined`.',
    );
    expect(
      checkPropTypes(
        { onClick: Card.propTypes.onClick },
        {},
        'prop',
        'Card',
      ),
    ).toBe(
      'Failed prop type: The prop `onClick` is marked as required in `Card`, but its value is `undefined`.',
    );
  });

  describe('when props are correct', () => {
    it('should renders the component correctly', () => {
      const { container } = render(
        <Card
          title='Title'
          text='Example text'
          onClick={onClick}
        />,
      );
      expect(container).toMatchSnapshot();
    });

    describe('and send button is clicked', () => {
      it('should call onClick', () => {
        render(
          <Card
            title='Title'
            text='Example text'
            onClick={onClick}
          />,
        );
        const sendButton = screen.getByText(/Enviar/);
        fireEvent.click(sendButton);
        expect(onClick).toBeCalled();
      });

      it('should hide title', () => {
        const { getByText } =render(
          <Card
            title='Title'
            text='Example text'
            onClick={onClick}
          />,
        );
        const sendButton = getByText(/Esconde titulo/);
        const title = getByText(/Title/);
        expect(title).toBeInTheDocument();

        act(() => {
          fireEvent.click(sendButton);
        });

        expect(title).not.toBeInTheDocument();
      });
    });
  });
});