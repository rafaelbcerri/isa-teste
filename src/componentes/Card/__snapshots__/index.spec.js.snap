// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card when props are correct should renders the component correctly 1`] = `
<div>
  <div
    class="card"
  >
    <h1>
      Title
    </h1>
    <hr />
    <p>
      Example text
    </p>
    <button>
      Enviar
    </button>
    <button>
      Esconde titulo
    </button>
  </div>
</div>
`;
